File: library-core/src/main/java/com/mikepenz/fastadapter/utils/EventHookUtil.java
package com.mikepenz.fastadapter.utils;

import android.support.v7.widget.RecyclerView;
import android.view.MotionEvent;
import android.view.View;

import com.mikepenz.fastadapter.FastAdapter;
import com.mikepenz.fastadapter.IItem;
import com.mikepenz.fastadapter.R;
import com.mikepenz.fastadapter.listeners.ClickEventHook;
import com.mikepenz.fastadapter.listeners.CustomEventHook;
import com.mikepenz.fastadapter.listeners.EventHook;
import com.mikepenz.fastadapter.listeners.LongClickEventHook;
import com.mikepenz.fastadapter.listeners.TouchEventHook;

import java.util.Arrays;
import java.util.List;

import javax.annotation.Nullable;

/**
 * Created by mikepenz on 25.01.16.
 */
public class EventHookUtil {
    /**
     * binds the hooks to the viewHolder
     *
     * @param viewHolder the viewHolder of the item
     */
    public static <Item extends IItem> void bind(RecyclerView.ViewHolder viewHolder, @Nullable final List<EventHook<Item>> eventHooks) {
        if (eventHooks == null) {
            return;
        }
        for (final EventHook<Item> event : eventHooks) {
            View view = event.onBind(viewHolder);
            if (view != null) {
                attachToView(event, viewHolder, view);
            }

            List<? extends View> views = event.onBindMany(viewHolder);
            if (views != null) {
                for (View v : views) {
                    attachToView(event, viewHolder, v);
                }
            }
        }
    }

    /**
     * attaches the specific event to a view
     *
     * @param event      the event to attach
     * @param viewHolder the viewHolder containing this view
     * @param view       the view to attach to
     */
    public static <Item extends IItem> void attachToView(final EventHook<Item> event, final RecyclerView.ViewHolder viewHolder, View view) {
        if (event instanceof ClickEventHook) {
            view.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    //get the adapter for this view
                    Object tagAdapter = viewHolder.itemView.getTag(R.id.fastadapter_item_adapter);
                    if (tagAdapter instanceof FastAdapter) {
                        FastAdapter<Item> adapter = (FastAdapter<Item>) tagAdapter;
                        //we get the adapterPosition from the viewHolder
                        int pos = adapter.getHolderAdapterPosition(viewHolder);
                        //make sure the click was done on a valid item
                        if (pos != RecyclerView.NO_POSITION) {
                            Item item = adapter.getItem(pos);
                            if (item != null) {
                                //we update our item with the changed property
                                ((ClickEventHook<Item>) event).onClick(v, pos, adapter, item);
                            }
                        }
                    }
                }
            });
        } else if (event instanceof LongClickEventHook) {
            view.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    //get the adapter for this view
                    Object tagAdapter = viewHolder.itemView.getTag(R.id.fastadapter_item_adapter);
                    if (tagAdapter instanceof FastAdapter) {
                        FastAdapter<Item> adapter = (FastAdapter<Item>) tagAdapter;
                        //we get the adapterPosition from the viewHolder
                        int pos = adapter.getHolderAdapterPosition(viewHolder);
                        //make sure the click was done on a valid item
                        if (pos != RecyclerView.NO_POSITION) {
                            Item item = adapter.getItem(pos);
                            if (item != null) {
                                //we update our item with the changed property
                                return ((LongClickEventHook<Item>) event).onLongClick(v, pos, adapter, item);
                            }
                        }
                    }
                    return false;
                }
            });
        } else if (event instanceof TouchEventHook) {
            view.setOnTouchListener(new View.OnTouchListener() {
                @Override
                public boolean onTouch(View v, MotionEvent e) {
                    //get the adapter for this view
                    Object tagAdapter = viewHolder.itemView.getTag(R.id.fastadapter_item_adapter);
                    if (tagAdapter instanceof FastAdapter) {
                        FastAdapter<Item> adapter = (FastAdapter<Item>) tagAdapter;
                        //we get the adapterPosition from the viewHolder
                        int pos = adapter.getHolderAdapterPosition(viewHolder);
                        //make sure the click was done on a valid item
                        if (pos != RecyclerView.NO_POSITION) {
                            Item item = adapter.getItem(pos);
                            if (item != null) {
                                //we update our item with the changed property
                                return ((TouchEventHook<Item>) event).onTouch(v, e, pos, adapter, item);
                            }
                        }
                    }
                    return false;
                }
            });
        } else if (event instanceof CustomEventHook) {
            //we trigger the event binding
            ((CustomEventHook<Item>) event).attachEvent(view, viewHolder);
        }
    }

    /**
     * a small helper to get a list of views from a dynamic amout of views.
     *
     * @param views the views to get as list
     * @return the list with the views
     */
    public static List<View> toList(View... views) {
        return Arrays.asList(views);
    }
}

File: library-core/src/main/java/com/mikepenz/fastadapter/utils/EventHookUtil.kt
package com.mikepenz.fastadapter.utils

import android.view.View
import androidx.recyclerview.widget.RecyclerView
import com.mikepenz.fastadapter.FastAdapter
import com.mikepenz.fastadapter.GenericItem
import com.mikepenz.fastadapter.R
import com.mikepenz.fastadapter.listeners.ClickEventHook
import com.mikepenz.fastadapter.listeners.CustomEventHook
import com.mikepenz.fastadapter.listeners.EventHook
import com.mikepenz.fastadapter.listeners.LongClickEventHook
import com.mikepenz.fastadapter.listeners.TouchEventHook

/**
 * Binds the hooks to the viewHolder
 *
 * @param viewHolder the viewHolder of the item
 */
internal fun List<EventHook<out GenericItem>>.bind(viewHolder: RecyclerView.ViewHolder) {
    for (event in this) {
        val view = event.onBind(viewHolder)
        if (view != null) {
            event.attachToView(viewHolder, view)
        }
        val views = event.onBindMany(viewHolder)
        if (views != null) {
            for (v in views) {
                event.attachToView(viewHolder, v)
            }
        }
    }
}

/**
 * Attaches the specific event to a view
 *
 * @param viewHolder the viewHolder containing this view
 * @param view       the view to attach to
 */
internal fun <Item : GenericItem> EventHook<Item>.attachToView(viewHolder: RecyclerView.ViewHolder, view: View) {
    when (this) {
        is ClickEventHook<*> -> view.setOnClickListener { v ->
            //get the adapter for this view
            val tagAdapter = viewHolder.itemView.getTag(R.id.fastadapter_item_adapter)
            val adapter = tagAdapter as? FastAdapter<Item> ?: return@setOnClickListener
            //we get the adapterPosition from the viewHolder
            val pos = adapter.getHolderAdapterPosition(viewHolder)
            //make sure the click was done on a valid item
            if (pos != RecyclerView.NO_POSITION) {
                val item: Item? = FastAdapter.getHolderAdapterItemTag(viewHolder)
                if (item != null) {
                    //we update our item with the changed property
                    (this as ClickEventHook<Item>).onClick(v, pos, adapter, item)
                }
            }
        }
        is LongClickEventHook<*> -> view.setOnLongClickListener(View.OnLongClickListener { v ->
            //get the adapter for this view
            val tagAdapter = viewHolder.itemView.getTag(R.id.fastadapter_item_adapter)
            val adapter = tagAdapter as? FastAdapter<Item> ?: return@OnLongClickListener false
            //we get the adapterPosition from the viewHolder
            val pos = adapter.getHolderAdapterPosition(viewHolder)
            //make sure the click was done on a valid item
            if (pos != RecyclerView.NO_POSITION) {
                val item: Item? = FastAdapter.getHolderAdapterItemTag(viewHolder)
                if (item != null) {
                    //we update our item with the changed property
                    return@OnLongClickListener (this as LongClickEventHook<Item>).onLongClick(v, pos, adapter, item)
                }
            }
            false
        })
        is TouchEventHook<*> -> view.setOnTouchListener(View.OnTouchListener { v, e ->
            //get the adapter for this view
            val tagAdapter = viewHolder.itemView.getTag(R.id.fastadapter_item_adapter)
            val adapter = tagAdapter as? FastAdapter<Item> ?: return@OnTouchListener false
            //we get the adapterPosition from the viewHolder
            val pos = adapter.getHolderAdapterPosition(viewHolder)
            //make sure the click was done on a valid item
            if (pos != RecyclerView.NO_POSITION) {
                val item: Item? = FastAdapter.getHolderAdapterItemTag(viewHolder)
                if (item != null) {
                    //we update our item with the changed property
                    return@OnTouchListener (this as TouchEventHook<Item>).onTouch(v, e, pos, adapter, item)
                }
            }
            false
        })
        //we trigger the event binding
        is CustomEventHook<*> -> (this as CustomEventHook<Item>).attachEvent(view, viewHolder)
    }
}
